glfw/xkb_glfw.c

```diff
@@ -308,11 +315,84 @@ glfw_xkb_update_x11_keyboard_id(_GLFWXKBData *xkb) {
     if (conn) state = xkb_x11_state_new_from_device(keymap, conn, xkb->keyboard_device_id); \
 }
 
+static void
+glfw_xkb_update_masks(_GLFWXKBData *xkb) {
+    bool succeeded = false;
+    unsigned used_bits = 0; /* To avoid using the same bit twice */
+    XkbDescPtr xkb_ptr = XkbGetMap( _glfw.x11.display, XkbVirtualModsMask | XkbVirtualModMapMask, XkbUseCoreKbd );
+
+#define S( a ) xkb->a##Idx = XKB_MOD_INVALID; xkb->a##Mask = 0
+    S(control); S(alt); S(shift); S(super); S(hyper); S(meta); S(capsLock); S(numLock);
+#undef S
+    if ( xkb_ptr ) {
+        Status status = XkbGetNames( _glfw.x11.display, XkbVirtualModNamesMask, xkb_ptr );
+        if ( status == Success ) {
+            for ( int indx = 0; indx < XkbNumVirtualMods; ++indx ) {
+                Atom atom = xkb_ptr->names->vmods[indx];
+                if ( atom ) {
+                    unsigned mask_rtn = 0;
+                    if ( XkbVirtualModsToReal( xkb_ptr, 1<<indx, &mask_rtn ) ) {
+                        const char *name = XGetAtomName( _glfw.x11.display, atom );
+#define S( a, s ) if ( !(used_bits & mask_rtn) && strcmp( name, #s ) == 0 ) xkb->a##Mask = mask_rtn, used_bits |= mask_rtn
+                        /* Note that the order matters here; earlier is higher priority. */
+                        S( alt, Alt );
+                        S( super, Super );
+                        S( numLock, NumLock );
+                        S( meta, Meta );
+                        S( hyper, Hyper );
+#undef S
+                    }
+                }
+            }
+            succeeded = true;
+        }
+        XkbFreeNames( xkb_ptr, XkbVirtualModNamesMask, True );
+        XkbFreeKeyboard( xkb_ptr, 0, True );
+    }
+    if ( succeeded ) {
+        unsigned indx, shifted;
+        for ( indx = 0, shifted = 1; used_bits; ++indx, shifted <<= 1, used_bits >>= 1 ) {
+#define S( a ) if ( xkb->a##Mask == shifted ) xkb->a##Idx = indx
+            S(alt); S(super); S(hyper); S(meta); S(numLock);
+#undef S
+        }
+    }
+#define S(a, n) xkb->a##Idx = xkb_keymap_mod_get_index(xkb->keymap, n); xkb->a##Mask = 1 << xkb->a##Idx;
+    S(control, XKB_MOD_NAME_CTRL);
+    S(shift, XKB_MOD_NAME_SHIFT);
+    S(capsLock, XKB_MOD_NAME_CAPS);
+    if ( !succeeded ) {
+        S(numLock, XKB_MOD_NAME_NUM);
+        S(alt, XKB_MOD_NAME_ALT);
+        S(super, XKB_MOD_NAME_LOGO);
+    }
+#undef S
+    debug( "Modifier indices alt:%d super:%d hyper:%d meta:%d num>ock:%d\n", xkb->altIdx, xkb->superIdx, xkb->hyperIdx, xkb->metaIdx, xkb->numLockIdx );
+}
+
+
 #else
 
 #define xkb_glfw_load_keymap(keymap, map_str) keymap = xkb_keymap_new_from_string(xkb->context, map_str, XKB_KEYMAP_FORMAT_TEXT_V1, 0);
 #define xkb_glfw_load_state(keymap, state) state = xkb_state_new(keymap);
 
+static void
+glfw_xkb_update_masks(_GLFWXKBData *xkb) {
```

Comment:
So hyper and meta work only under X11, correct?

Reply:
Yes, X11 with XKB. I can provide a patch that I think Should Work (TM) for non-X11 XKB by looking at modifier names provided by libxkbcommon (using the same technique as in the patch), but I have no way of validating such a patch.